src/peft/tuners/reft/layer.py

```diff
@@ -0,0 +1,262 @@
+
+        # Create weights with provided shape
+        self.create_adapter_parameters(adapter_name, r, shape)
+
+        # Initialize weights
+        if init_weights:
+            self.reset_adapter_parameters(adapter_name)
+
+        # Move new weights to device
+        weight = getattr(self.get_base_layer(), "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                self.to(weight.device, dtype=weight.dtype)
+            else:
+                self.to(weight.device)
+        self.set_adapter(self.active_adapters)
+
+    def get_delta_weight(self, adapter_name: str) -> torch.Tensor:
+        return self.reft_R[active_adapter]
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            result = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            result = self.base_layer(x, *args, **kwargs)
+        else:
+            result = self.base_layer(x, *args, **kwargs)
+
+            # Execute all the adapters
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self._available_adapters:
+                    continue
+
+                rotate_layer = self.reft_R[active_adapter]
+                learned_source = self.reft_A[active_adapter]
+                module_dropout = self.module_dropout[active_adapter]
+                if self.loc[active_adapter] is None:
+                    rotated_base = rotate_layer(result)
+                    offset = torch.matmul((learned_source(result) - rotated_base), rotate_layer.weight.T)
+                    output = result + offset
+                else:
+                    loc = self.loc[active_adapter]
+                    selected_results = tensor.gather(result, 1, loc)
+                    rotated_base = rotate_layer(selected_results)
+                    offset = torch.matmul((learned_source(selected_results) - rotated_base), rotate_layer.weight.T)
+                    tensor.scatter_(output, 1, loc, offset)
+
+        output = output.to(previous_dtype)
+        return output
+
+
+class Linear(ReftLayer):
+    """Reft implemented in Linear layer"""
+
+    def __init__(
+        self,
+        base_layer: nn.Module,
+        adapter_name: str = "default",
+        r: int = 0,
+        alpha: float = 0.0,
+        rank_dropout: float = 0.0,
+        module_dropout: float = 0.0,
+        init_weights: bool = True,
+        **kwargs,
+    ):
+        super().__init__(base_layer)
+
+        # Create adapter and set it active
+        self._active_adapter = adapter_name
+        self.update_layer(adapter_name, r, alpha, rank_dropout, module_dropout, init_weights, **kwargs)
+
+    def _get_delta_activations(
```

Comment:
Is this used anywhere?

Reply:
This is not used. Since LycorisTuner defines `_get_delta_activations` as `abstractmethod`, I define  `_get_delta_activations` to avoid raising errors.

Reply:
I see. In that case, it should probably not be an abstract method on `LycorsLayer` to begin with, but let's not bother with this here. I'd just raise a `NotImplementedError` then, instead of having dead code.